pkg/metrics/node.go

```diff
@@ -0,0 +1,194 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rancher/wrangler/pkg/ticker"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	mgmtcontrollers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+)
+
+const (
+	nodeProviderLabel  = "provider"
+	nodeClusterIDLabel = "cluster_id"
+)
+
+var (
+	nodeLabels = []string{nodeProviderLabel, nodeClusterIDLabel}
+	numNodes   = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "nodes",
+			Help:      "Number of nodes in ready clusters",
+		}, nodeLabels,
+	)
+	numCores = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "node_cores",
+			Help:      "Number of node cores in ready clusters",
+		}, nodeLabels,
+	)
+)
+
+const (
+	nodePollInterval = time.Duration(15 * time.Second)
```

Comment:
I considered making the node poll interval a setting, but that I opted not to for a few reasons:
1. The node count and core counts won't change all too often. Scaling clusters, creating new clusters, removing clusters, etc are the operations that would cause these metric values to change
2. The time series functionality for this metric, i.e. seeing how it changed over time is less important than the point in time value
3. User would have control if it's a setting and nothing stopping them from setting it to a very large value

The metrics don't need to be available right away, but they do need to update often enough to give a clear picture of node usage. So for instance we shouldn't wait minutes / hours before returning accurate node and node core count metrics

Reply:
It makes sense to not allow messing with this setting. I was more thinking that we would want to allow users to control the rate so that we don't end up causing usage issues with the endpoints that it is using. But this is infrequent enough to not make an impact, and if we need them at a certain rate it makes sense to not allow adjustments.